server/idapiProxy.ts

```diff
@@ -0,0 +1,180 @@
+});
+
+const prepareBody = <T>(body: T | undefined) => {
+	if (!body) {
+		return undefined;
+	} else if (typeof body === 'string') {
+		return body;
+	}
+	try {
+		return JSON.stringify(body);
+	} catch (e) {
+		throw new Error(`Error stringifying request body: ${e}`);
+	}
+};
+
+export const setOptions = ({
+	path,
+	subdomain,
+	method,
+	cookies,
+	config,
+}: {
+	path: string;
+	subdomain: string;
+	method: HTTPMethod;
+	cookies: CookiesWithToken;
+	config: IdapiConfig;
+}): IdapiFetchOptions => {
+	const hostname = `${subdomain}.${getBaseDomain()}`;
+
+	const headers = {
+		'X-GU-ID-Client-Access-Token': `Bearer ${config.accessToken}`,
+		...securityCookieToHeader(cookies),
+		'Content-Type': 'application/json',
+		Origin: `https://manage.${getBaseDomain()}`,
+		Referer: `https://manage.${getBaseDomain()}`,
+		// Avatar API expects a Cookie header with the SC_GU_U cookie.
+		Cookie: subdomain === 'avatar' ? `SC_GU_U=${cookies.SC_GU_U};` : '',
+	};
+
+	const options = {
+		headers,
+		method,
+		route: `https://${hostname}${path}`,
+	};
+
+	return options;
+};
+
+/**
+ * Performs a fetch request to IDAPI.
+
+ * @template T - The type of the JSON payload sent to IDAPI
+ */
+export const idapiFetch = async <T>({
+	options,
+	body,
+}: {
+	options: IdapiFetchOptions;
+	body?: T;
+}) => {
+	const response = await fetch(options.route, {
+		method: options.method,
+		headers: options.headers,
+		body: prepareBody<T>(body),
+	});
+	return response;
+};
+
+/**
+ * Route handler which proxies a request made from the MMA client to IDAPI.
+ *
+ * @template T The type of the JSON response from IDAPI
+ * @param url The IDAPI endpoint to hit
+ * @param method The HTTP method to use (default: 'GET')
+ * @param processData A function to process the JSON response (optional)
+ * @returns An Express route handler
+ */
+export const idapiProxyHandler =
+	<T>({
+		url,
+		method = 'GET',
+		processData,
+	}: {
+		url: string;
+		method?: HTTPMethod;
+		processData?: (json: T, res: Response) => Response | void;
+	}) =>
+	async (req: Request, res: Response, next: NextFunction) => {
+		let config;
+		try {
+			config = await getConfig();
+		} catch (e) {
+			return handleError(e, res, next);
+		}
+		const options = setOptions({
+			path: url,
+			subdomain: 'idapi',
+			method,
+			cookies: req.cookies,
+			config,
+		});
+		try {
+			const response = await idapiFetch<T>({
+				options,
+				body: ['POST', 'PUT', 'PATCH'].includes(method)
+					? req.body
+					: undefined,
+			});
+			if (response.status === 204) {
+				return res.sendStatus(204);
+			} else {
+				try {
+					const json = await response.json();
+					if (processData) {
+						return processData(json, res);
+					} else {
+						// Just send the JSON response as is
+						return res.status(response.status).json(json);
+					}
+				} catch (e) {
+					// Swallow JSON parse errors
+					if (e instanceof SyntaxError) {
+						return res.sendStatus(response.status);
```

Comment:
Because IDAPI (or any API, really) might send us an error response with an error status code, but no valid JSON body. If we can't parse the body, we can at least send the status code. It's not really an error that's worth throwing. If there was a way to check that a body _is_ JSON _without_ parsing it and potentially throwing an error, we'd do that.

Reply:
👍